crates/optimism/txpool/src/supervisor/errors.rs

```diff
@@ -7,17 +94,237 @@ pub enum InteropTxValidatorError {
+
+                    // UNIMPLEMENTED errors
+                    SupervisorErrorCode::CANNOT_GET_PARENT_OF_FIRST_BLOCK_IN_DB => {
+                        return Self::InvalidEntry(InvalidInboxEntry::NoParentForFirstBlock);
+                    }
+
+                    // UNAVAILABLE errors
+                    SupervisorErrorCode::FUTURE_DATA => {
+                        return Self::InvalidEntry(InvalidInboxEntry::FutureData);
+                    }
+
+                    // DATA_LOSS errors
+                    SupervisorErrorCode::MISSED_DATA => {
+                        return Self::InvalidEntry(InvalidInboxEntry::MissedData);
+                    }
+                    SupervisorErrorCode::DATA_CORRUPTION => {
+                        return Self::InvalidEntry(InvalidInboxEntry::DataCorruption);
+                    }
+                }
+            }
+        }
+
+        // Default to generic error
         Self::Other(Box::new(err))
     }
 }
+
+/// Extracts a chain ID from error messages like "unsupported chain id: 1234"
+fn extract_chain_id(message: &str) -> Option<u64> {
+    // Common pattern for chain ID errors
+    if let Some(idx) = message.find("chain id") {
+        if idx + 8 < message.len() {
+            // Ensure there are characters after "chain id"
+            let remainder = &message[idx + 8..];
+            if let Some(colon_idx) = remainder.find(':') {
+                #[allow(clippy::int_plus_one)]
+                if colon_idx + 1 <= remainder.len() {
+                    // Ensure there are characters after the colon
+                    let num_part = &remainder[colon_idx + 1..].trim();
+                    return num_part.parse::<u64>().ok();
+                }
+            }
+        }
+    }
+
+    // Fallback: look for any number in the message
+    #[allow(clippy::is_digit_ascii_radix)]
+    message
+        .split_whitespace()
+        .filter_map(|word| word.trim_matches(|c: char| !c.is_digit(10)).parse::<u64>().ok())
+        .next()
+}
+
+/// Extracts a safety level value from error message patterns
+fn extract_safety_level(message: &str, key_param: &str) -> Option<SafetyLevel> {
+    // Look for patterns like "got: 0" or "expected: 1" in the message
+    let patterns = [format!("{key_param}: "), format!("{key_param} ")];
+    for pattern in &patterns {
+        if let Some(pos) = message.find(pattern.as_str()) {
+            if pos + pattern.len() < message.len() {
+                // Ensure there are characters after the pattern
+                let value_part = &message[pos + pattern.len()..];
+                let value = value_part.chars().next().and_then(|c| c.to_digit(10));
+
+                return match value {
+                    Some(0) => Some(SafetyLevel::Unsafe),
+                    Some(1) => Some(SafetyLevel::Safe),
+                    Some(2) => Some(SafetyLevel::Finalized),
+                    _ => None,
+                };
+            }
+        }
+    }
+    None
+}
```

Comment:
let's skip this, and not parse the rpc error message at all and just rely on the code. error code is less risk for bugs than string error messages which may be worded differently across implementations. differing wording in strings can also be a harder bug to spot than spotting if the wrong error code is used across implementations, the latter is fairly straight forward bug to spot.